ocaml/jane/doc/proposals/kind-inference.md

```diff
@@ -0,0 +1,827 @@
+tconstr_jkind = π [[ 'aᵢ : κᵢ ]]. κ₀
+------------------------------------------------- D_KINDED
+Γ ⊢ type type_params t : jkind ↠ t : tconstr_jkind
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ τ : κ₀
+Γ' = t = λ [[ 'aᵢ : κᵢ ]]. τ
+--------------------------------------------- D_ABBREV
+Γ ⊢ type type_params t = τ ↠ Γ'
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ τ : κ₀
+Γ' = t : π [[ 'aᵢ : κᵢ ]]. κ₀   (* private: make [t] abstract in the result *)
+---------------------------------------- D_PRIVATE
+Γ ⊢ type type_params t = private τ ↠ Γ'
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+Γ, [[ 'aᵢ : κᵢ ]], t : π [[ 'aᵢ : κᵢ ]]. κ₀ ⊢tk type_kind ↠ δ : κ₀
+--------------------------------------------- D_NOMINATIVE
+Γ ⊢ type type_params t = type_kind ↠ t := λ [[ 'aᵢ : κᵢ ]]. δ : κ₀
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ τ : κ₀
+Γ' = t = λ [[ 'aᵢ : κᵢ ]]. τ
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ jkind ↠ κ₀'
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ κ₀ ≤ κ₀'
+------------------------------------- D_KINDED_ABBREV
+Γ ⊢ type type_params t : jkind = τ ↠ Γ'
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ τ : κ₀
+Γ' = t : π [[ 'aᵢ : κᵢ ]]. κ₀'        (* κ₀', not κ₀, for private abbrevs *)
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ jkind ↠ κ₀'
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ κ₀ ≤ κ₀'
+--------------------------------------------- D_KINDED_PRIVATE
+Γ ⊢ type type_params t : jkind = private τ ↠ Γ'
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+τ = [[ 'aᵢ ]] t'
+Γ, [[ 'aᵢ : κᵢ ]], t : π [[ 'aᵢ : κᵢ ]]. κ₀ ⊢ τ : κ₀
+Γ' = t = λ [[ 'aᵢ : κᵢ ]]. τ
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢tk type_kind ↠ δ : κ₀'
+t' := λ [[ 'aᵢ : κᵢ ]]. δ' : κ₀'' ∈ Γ
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ δ reexports δ'
+----------------------------------------- D_REEXPORT
+Γ ⊢ type type_params t = τ = type_kind ↠ Γ'
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+Γ, [[ 'aᵢ : κᵢ ]], t : π [[ 'aᵢ : κᵢ ]]. κ₀ ⊢tk type_kind ↠ δ : κ₀'
+Γ' = t := λ [[ 'aᵢ : κᵢ ]]. δ : κ₀  (* user-written jkind is preserved *)
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ jkind ↠ κ₀
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ κ₀' ≤ κ₀
+----------------------------------------------- D_KINDED_NOMINATIVE
+Γ ⊢ type type_params t : jkind = type_kind ↠ Γ'
+
+Γ ⊢ type_params ↠ [[ 'aᵢ : κᵢ ]]
+τ = [[ 'aᵢ ]] t'
+Γ, [[ 'aᵢ : κᵢ ]], t : π [[ 'aᵢ : κᵢ ]]. κ₀ ⊢ τ : κ₀
+Γ' = t = λ [[ 'aᵢ : κᵢ ]]. τ
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢tk type_kind ↠ δ : κ₀'
+t' := λ [[ 'aᵢ : κᵢ ]]. δ' ∈ Γ
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ δ reexports δ'
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ jkind ↠ κ₀''
+Γ, [[ 'aᵢ : κᵢ ]], Γ' ⊢ κ₀ ≤ κ₀''
+-------------------------------------------------- D_KINDED_REEXPORT
+Γ ⊢ type type_params t : jkind = τ = type_kind ↠ Γ'
+
+
+Γ ⊢find t ∈ Γ₂ ↠ tconstr_jkind  (* find [t] in Γ₂ *)
+======================================
+  (* pre-condition: Γ₂ ⊆ Γ *)
+
+t : tconstr_jkind ∈ Γ₂
+--------------------------- FIND_ABSTRACT
+Γ ⊢find t ∈ Γ₂ ↠ tconstr_jkind
+
+t = λ [[ 'aᵢ : κᵢ ]]. τ ∈ Γ₂
+Γ, [[ 'aᵢ : κᵢ ]] ⊢ τ : κ₀
+------------------------------------- FIND_ABBREV
+Γ ⊢find t ∈ Γ₂ ↠ π [[ 'aᵢ : κᵢ ]]. κ₀
```

Comment:
This is wrong. It should use `kind_of`. Otherwise

```ocaml
module M : sig
  type ('a : value) t : value
end = struct
  type ('a : any) t = 'a
end
```

is rejected, because the struct kind is just `any`. It needs to be `kind_of 'a`, in which case all would be well.